MdeModulePkg/Library/MemoryBucketLib/MemoryBucketLib.c

```diff
@@ -0,0 +1,447 @@
+
+  CurrentBucketTops[Index] = NewTop;
+}
+
+/**
+  This function gets the address of the startfree memory in the bucket
+  specified by MemoryType.
+
+  @param[in] MemoryType   The type of memory we are interested in.
+
+  @retval    Returns an address that points to the start of free memory in a
+             memory bucket that is specified by MemoryType
+**/
+EFI_PHYSICAL_ADDRESS
+EFIAPI
+GetCurrentBucketTop (
+  EFI_MEMORY_TYPE  MemoryType
+  )
+{
+  UINTN  Index;
+
+  Index = MemoryTypeToIndex (MemoryType);
+
+  return CurrentBucketTops[Index];
+}
+
+/**
+  This function gets the address of the bottom of the bucket specified by
+  MemoryType.
+
+  @param[in] MemoryType   The type of memory we are interested in.
+
+  @retval    Returns an address that points to the bottom of memory in a
+             memory bucket that is specified by MemoryType
+**/
+EFI_PHYSICAL_ADDRESS
+EFIAPI
+GetCurrentBucketBottom (
+  EFI_MEMORY_TYPE  MemoryType
+  )
+{
+  EFI_PHYSICAL_ADDRESS  ReturnValue;
+  UINTN                 Index;
+
+  Index = MemoryTypeToIndex (MemoryType);
+
+  ReturnValue = (EFI_PHYSICAL_ADDRESS)(RuntimeMemoryStats[Index].BaseAddress) - (EFI_PAGE_SIZE * RuntimeMemoryStats[Index].NumberOfPages);
+  return ReturnValue;
+}
+
+/**
+  Function that returns the address associated with the end of the
+  memory bucket structure.
+
+  @retval    The memory address below the memory bucket structure.
+
+**/
+EFI_PHYSICAL_ADDRESS
+EFIAPI
+GetEndOfBucketsAddress (
+  VOID
+  )
+{
+  UINTN  Index;
+
+  for (Index = NumberOfBuckets-1; Index >= 0; Index--) {
+    if (RuntimeMemoryStats[Index].NumberOfPages > 0) {
+      return GetCurrentBucketBottom (MemoryTypes[Index]);
+    }
+  }
+
+  DEBUG ((DEBUG_ERROR, "We should not be calling GetEndOfBucketsAddress if we have Pei memory buckets disabled!\n"));
+  return 0;
+}
+
+/**
+  This function updates the memory bucket structure to include a new memory allocation.
+
+  @param[in] Pages        The number of pages we are allocating in the bucket.
+  @param[in] MemoryType   The type of memory we are updating in the memory bucket
+                          structure.
+**/
+VOID
+EFIAPI
+UpdateRuntimeMemoryStats (
+  UINTN            Pages,
+  EFI_MEMORY_TYPE  MemoryType
+  )
+{
+  UINTN  Index;
+
+  Index = MemoryTypeToIndex (MemoryType);
+
+  if (RuntimeMemoryStats[Index].CurrentNumberOfPages + (UINT64)Pages > RuntimeMemoryStats[Index].NumberOfPages) {
+    DEBUG ((DEBUG_ERROR, "We have overflowed while allocating PEI pages of index: %d!\n", Index));
+    ASSERT (FALSE);
+    return;
+  }
+
+  RuntimeMemoryStats[Index].CurrentNumberOfPages = (UINT64)RuntimeMemoryStats[Index].CurrentNumberOfPages + (UINT64)Pages;
+
+  RuntimeBucketHob.RuntimeBuckets[Index]     = RuntimeMemoryStats[Index];
+  RuntimeBucketHob.CurrentTopInBucket[Index] = CurrentBucketTops[Index];
+}
+
+/**
+  This function checks to see if there is an attempt to allocate memory in the runtime
+  memory buckets that we have defined.

```

Comment:
The current description makes it sound like this is checking the actual memory allocation after the fact. But this function is really checking an address likely before the allocation.

```suggestion
  This function checks to see if the given address lies within the currently defined
  runtime memory bucket range.
```